File: app/src/main/java/jp/juggler/subwaytooter/action/Action_Notification.java
package jp.juggler.subwaytooter.action;

import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;

import jp.juggler.subwaytooter.ActMain;
import jp.juggler.subwaytooter.App1;
import jp.juggler.subwaytooter.Column;
import jp.juggler.subwaytooter.R;
import jp.juggler.subwaytooter.api.TootApiClient;
import jp.juggler.subwaytooter.api.TootApiResult;
import jp.juggler.subwaytooter.api.TootApiTask;
import jp.juggler.subwaytooter.api.entity.TootNotification;
import jp.juggler.subwaytooter.table.SavedAccount;
import jp.juggler.subwaytooter.util.Utils;
import okhttp3.Request;
import okhttp3.RequestBody;

@SuppressWarnings("WeakerAccess") @SuppressLint("StaticFieldLeak")
public class Action_Notification {
	
	public static void deleteAll(
		@NonNull final ActMain activity
		, @NonNull final SavedAccount target_account
		, boolean bConfirmed
	){
		if( ! bConfirmed ){
			new AlertDialog.Builder( activity )
				.setMessage( R.string.confirm_delete_notification )
				.setNegativeButton( R.string.cancel, null )
				.setPositiveButton( R.string.ok, new DialogInterface.OnClickListener() {
					@Override
					public void onClick( DialogInterface dialog, int which ){
						deleteAll( activity, target_account, true );
					}
				} )
				.show();
			return;
		}
		new TootApiTask( activity, target_account, true ) {
			
			@Override protected TootApiResult doInBackground( Void... params ){
				
				Request.Builder request_builder = new Request.Builder().post(
					RequestBody.create(
						TootApiClient.MEDIA_TYPE_FORM_URL_ENCODED
						, "" // 空データ
					) );
				return client.request( "/api/v1/notifications/clear", request_builder );
			}
			
			@Override protected void handleResult( TootApiResult result ){
				if( result == null ) return; // cancelled.
				
				if( result.object != null ){
					// ok. api have return empty object.
					for( Column column : App1.app_state.column_list ){
						if( column.column_type == Column.TYPE_NOTIFICATIONS
							&& column.access_info.acct.equals( target_account.acct )
							){
							column.removeNotifications();
						}
					}
					Utils.showToast( activity, false, R.string.delete_succeeded );
				}else{
					Utils.showToast( activity, false, result.error );
				}
			}
			
		}.executeOnExecutor( App1.task_executor );
	}
	
	public static void deleteOne(
		@NonNull final ActMain activity
		, @NonNull final SavedAccount access_info
		, @NonNull final TootNotification notification
	){
		new TootApiTask( activity, access_info, true ) {
			
			@Override protected TootApiResult doInBackground( Void... params ){
				
				Request.Builder request_builder = new Request.Builder()
					.post( RequestBody.create( TootApiClient.MEDIA_TYPE_FORM_URL_ENCODED
						, "id=" + Long.toString( notification.id )
						)
					);
				
				return client.request(
					"/api/v1/notifications/dismiss"
					, request_builder );
			}
			
			@Override protected void handleResult( TootApiResult result ){
				
				if( result == null ){
					// cancelled.
				}else if( result.object != null ){
					// 成功したら空オブジェクトが返される
					for( Column column : App1.app_state.column_list ){
						column.removeNotificationOne( access_info, notification );
					}
					Utils.showToast( activity, true, R.string.delete_succeeded );
				}else{
					Utils.showToast( activity, true, result.error );
				}
			}
		}.executeOnExecutor( App1.task_executor );
	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/action/Action_Notification.kt
package jp.juggler.subwaytooter.action

import androidx.appcompat.app.AlertDialog
import jp.juggler.subwaytooter.ActMain
import jp.juggler.subwaytooter.R
import jp.juggler.subwaytooter.actmain.addColumn
import jp.juggler.subwaytooter.api.entity.TootAccount
import jp.juggler.subwaytooter.api.entity.TootNotification
import jp.juggler.subwaytooter.api.runApiTask
import jp.juggler.subwaytooter.column.ColumnType
import jp.juggler.subwaytooter.column.isNotificationColumn
import jp.juggler.subwaytooter.column.removeNotificationOne
import jp.juggler.subwaytooter.column.removeNotifications
import jp.juggler.subwaytooter.table.SavedAccount
import jp.juggler.util.coroutine.launchMain
import jp.juggler.util.log.showToast
import jp.juggler.util.network.toFormRequestBody
import jp.juggler.util.network.toPost

fun ActMain.clickNotificationFrom(
    pos: Int,
    accessInfo: SavedAccount,
    who: TootAccount,
) {
    if (accessInfo.isMisskey) {
        showToast(false, R.string.misskey_account_not_supported)
    } else {
        accessInfo.getFullAcct(who).validFull()?.let {
            addColumn(
                pos,
                accessInfo,
                ColumnType.NOTIFICATION_FROM_ACCT,
                it
            )
        }
    }
}

fun ActMain.notificationDeleteOne(
    accessInfo: SavedAccount,
    notification: TootNotification?,
) {
    notification ?: return

    launchMain {
        runApiTask(accessInfo) { client ->
            // https://github.com/tootsuite/mastodon/commit/30f5bcf3e749be9651ed39a07b893f70605f8a39
            // 2種類のAPIがあり、片方は除去された

            // まず新しいAPIを試す
            val result = client.request(
                "/api/v1/notifications/${notification.id}/dismiss",
                "".toFormRequestBody().toPost()
            )

            when (result?.response?.code) {

                // 新しいAPIがない場合、古いAPIを試す
                422 -> client.request(
                    "/api/v1/notifications/dismiss",
                    "id=${notification.id}".toFormRequestBody().toPost()
                )

                else -> result
            }
        }?.let { result ->
            when (result.jsonObject) {
                null -> showToast(true, result.error)
                else -> {
                    // 成功したら空オブジェクトが返される
                    for (column in appState.columnList) {
                        column.removeNotificationOne(accessInfo, notification)
                    }
                    showToast(true, R.string.delete_succeeded)
                }
            }
        }
    }
}

fun ActMain.notificationDeleteAll(
    targetAccount: SavedAccount,
    bConfirmed: Boolean = false,
) {
    if (!bConfirmed) {
        AlertDialog.Builder(this)
            .setMessage(R.string.confirm_delete_notification)
            .setNegativeButton(R.string.cancel, null)
            .setPositiveButton(R.string.ok) { _, _ ->
                notificationDeleteAll(targetAccount, true)
            }
            .show()
        return
    }

    launchMain {
        runApiTask(targetAccount) { client ->
            client.request(
                "/api/v1/notifications/clear",
                "".toFormRequestBody().toPost()
            )
        }?.let { result ->
            when (result.jsonObject) {
                null -> showToast(false, result.error)

                else -> {
                    // ok. api have return empty object.
                    for (column in appState.columnList) {
                        if (column.isNotificationColumn && column.accessInfo == targetAccount) {
                            column.removeNotifications()
                        }
                    }
                    showToast(false, R.string.delete_succeeded)
                }
            }
        }
    }
}
